### PR TITLE
set default subtitle from hls

### LINKED ISSUE #687: [FEATURE] HLS don't start with or give access to know which subtitle is the default
**Is your feature request related to a problem? Please describe.**
When I use HLS (eg. m3u8) I receive an attribute when a subtitle is the default the `DEFAULT=YES`, but I can't access this field from the lib and the player don't start with the subtitle with this field marked.

**Describe the solution you'd like**
I wait that the video start with the subtitle that have the attribute `DEFAULT=YES`, a used pattern in the players, or at least give the information of which subtitle is the default.

**Additional context**
I saw another feature requested [[FEATURE] Enable Default Subtitle](https://github.com/jhomlala/betterplayer/issues/615) , but my request differs because the HLS came ready to play and not is needed set the subtitles.

The example project have a video that can be used, the `HLS subtitles`, if you play in the most of players the default sub will be selected by default.
In the file of [example](https://bitdash-a.akamaihd.net/content/sintel/hls/playlist.m3u8) you can see this line below with the field `DEFAULT=YES`
`#EXT-X-MEDIA:TYPE=SUBTITLES,GROUP-ID="subs",NAME="English",DEFAULT=YES,AUTOSELECT=YES,FORCED=NO,LANGUAGE="en",URI="subtitles_en.m3u8"`

### PR DESCRIPTION
**Pull request by siloebb**
fix #687
This pull request make HLS video set the default subtitle using `DEFAULT=YES`, so when is set the subtitle will start by default.

### CODE REVIEW

File: lib/src/hls/better_player_hls_utils.dart [hunk 1]
```diff
@@ -137,6 +138,16 @@ class BetterPlayerHlsUtils {
         targetDuration = parsedSubtitle.targetDurationUs! ~/ 1000;
       }
 
+      bool isDefault = false;
+      // bool isForced = false;
```
Comment (jhomlala) [inline lib/src/hls/better_player_hls_utils.dart:142]: Please remove these leftovers.

File: lib/src/hls/better_player_hls_utils.dart [hunk 2]
```diff
@@ -137,6 +138,16 @@ class BetterPlayerHlsUtils {
...
+      // bool isAutoSelect = false;
+
+      if(rendition.format.selectionFlags != null) {
+        isDefault = Util.checkBitPositionIsSet(rendition.format.selectionFlags!, 1);
+        // isForced = Util.checkBitPositionIsSet(rendition.format.selectionFlags!, 2);
```
Comment (jhomlala) [inline lib/src/hls/better_player_hls_utils.dart:147]: Please remove these leftovers.

File: lib/src/hls/hls_parser/format.dart [hunk 3]
```diff
@@ -37,7 +38,8 @@ class Format {
           required int? height,
           required double? frameRate,
           int selectionFlags = Util.selectionFlagDefault,
-          int? roleFlags}) =>
+          int? roleFlags,
+          bool? isDefault}) =>
```
Comment (jhomlala) [inline lib/src/hls/hls_parser/format.dart:42]: Please add `,` at the end.

File: lib/src/hls/better_player_hls_utils.dart [hunk 4]
```diff
@@ -137,6 +138,16 @@ class BetterPlayerHlsUtils {
         targetDuration = parsedSubtitle.targetDurationUs! ~/ 1000;
       }
 
+      bool isDefault = false;
+      // bool isForced = false;
+      // bool isAutoSelect = false;
+
+      if(rendition.format.selectionFlags != null) {
+        isDefault = Util.checkBitPositionIsSet(rendition.format.selectionFlags!, 1);
```
Comment (jhomlala) [inline lib/src/hls/better_player_hls_utils.dart:146]: Why do you need to check that with bits?

File: lib/src/hls/hls_parser/hls_playlist_parser.dart [hunk 5]
```diff
@@ -622,8 +623,12 @@ class HlsPlaylistParser {
     required String pattern,
     required bool defaultValue,
   }) {
-    final List<Match> list = line.allMatches(pattern).toList();
-    return list.isEmpty ? defaultValue : list.first.pattern == booleanTrue;
+    final regExp = RegExp(pattern);
```
Comment (jhomlala) [inline lib/src/hls/hls_parser/hls_playlist_parser.dart:626]: Why did you change this code?

File: lib/src/hls/better_player_hls_utils.dart (hunk 4 above)
Comment (siloebb) [reply lib/src/hls/better_player_hls_utils.dart:146]: I just used the pattern that was being used, I try not to change how the parser work.

You can see in file `util.dart`

```
class Util {
  static const int selectionFlagDefault = 1;
  static const int selectionFlagForced = 1 << 1; // 2
  static const int selectionFlagAutoSelect = 1 << 2; // 4
...
}
```
and in the file `hls_playlist_parser.dart`

```
...
if (parseOptionalBooleanAttribute(
        line: line,
        pattern: regexpDefault,
        defaultValue: false)) flags |= Util.selectionFlagDefault;
...
```

these are operation with bits.

File: lib/src/hls/hls_parser/hls_playlist_parser.dart (hunk 5 above)
Comment (siloebb) [reply lib/src/hls/hls_parser/hls_playlist_parser.dart:626]: Using String can't find the pattern, I don't know why, so I need to use Regex to find.

I leave a code that can be used in Dartcode to show that just don't work as expected.

```
void main() { 
  final line = '#EXT-X-MEDIA:TYPE=SUBTITLES,GROUP-ID="subs",NAME="English",DEFAULT=YES,AUTOSELECT=YES,FORCED=NO,LANGUAGE="en",URI="subtitles_en.m3u8"';
  final pattern = 'DEFAULT=(YES|NO)';
  
  print('Found with Regex first: ${useRegexToFind(line, pattern)}');
  print('Found with Regex From String: ${useRegexInList(line, pattern)}');  
}

bool useRegexToFind(String line, String patternString){
  final pattern = RegExp(patternString);  
  final List<Match> list = pattern.allMatches(line).toList();
  
  return list.isEmpty; //If true, dont find
}

bool useRegexInList(String line, String pattern){
  final List<Match> list = line.allMatches(pattern).toList();
  return list.isEmpty; //If true, dont find
}

```
I leave the results too.

Captura de Tela 2021-09-13 às 09 13 59

**Comment by siloebb** [comment]
I made another PR removing the unecessary comments.

File: lib/src/hls/better_player_hls_utils.dart (hunk 1 above)
Comment (siloebb) [reply lib/src/hls/better_player_hls_utils.dart]: ok

File: lib/src/hls/better_player_hls_utils.dart (hunk 2 above)
Comment (siloebb) [reply lib/src/hls/better_player_hls_utils.dart]: ok

File: lib/src/hls/hls_parser/format.dart (hunk 3 above)
Comment (siloebb) [reply lib/src/hls/hls_parser/format.dart]: ok

**Comment by jhomlala** [comment]
LGTM thanks.
